File: native-benchmarks/native-benchmarks/camera/CameraBenchmark.swift
//
//  CameraBenchmark.swift
//  native-benchmarks
//
//  Created by Vanja Vidmark on 2025-04-23.
//

import Foundation

final class CameraBenchmark {
    private let cameraService = CameraService()
    private let performanceCalculator: HardwarePerformanceCalculator

    init(performanceCalculator: HardwarePerformanceCalculator) {
        self.performanceCalculator = performanceCalculator
    }

    func runBenchmark(warmup: Int, n: Int, measureTime: Bool) async {
        do {
            try await cameraService.prepare()
        } catch {
            print("Camera setup failed: \(error)")
            return
        }

        for i in 0..<warmup {
            // warmup round
            try? await cameraService.takeAndSavePhoto()
            print("Saved warmup photo \(i + 1)/\(warmup)")
        }
        if measureTime {
            for i in 0..<n {
                let start = ProcessInfo.processInfo.systemUptime
                try? await cameraService.takeAndSavePhoto()
                let timeElapsed = ProcessInfo.processInfo.systemUptime - start
                performanceCalculator.sampleTime(duration: timeElapsed)
                print("Saved photo \(i + 1)/\(n)")
            }
            performanceCalculator.postTimes()
        } else {
            performanceCalculator.start()
            for i in 0..<n {
                performanceCalculator.markIteration(i)
                try? await cameraService.takeAndSavePhoto()
                print("Saved photo \(i + 1)/\(n)")
            }
            performanceCalculator.stopAndPost()
        }
        cameraService.stop()
        print("Camera benchmark complete")
    }
}

File: kmp-benchmarks/composeApp/src/commonMain/kotlin/org/example/kmpbenchmarks/camera/CameraBenchmark.kt
package org.example.kmpbenchmarks.camera

import org.example.kmpbenchmarks.PerformanceCalculator
import org.example.kmpbenchmarks.getTime

expect suspend fun prepareCameraSession()
expect suspend fun takeAndSavePhoto()
expect fun stopCameraSession()

class CameraBenchmark(
    private val performanceCalculator: PerformanceCalculator
) {
    suspend fun runBenchmark(warmup: Int, n: Int, measureTime: Boolean) {
        try {
            prepareCameraSession()
        } catch (e: Exception) {
            println("Camera setup failed: ${e.message}")
            return
        }

        // Warmup round
        for (i in 0 until warmup) {
            try {
                takeAndSavePhoto()
                println("Saved warmup photo ${i + 1}/$warmup")
            } catch (e: Exception) {
                println("Warmup photo $i failed: ${e.message}")
            }
        }

        if (measureTime) {
            for (i in 0 until n) {
                val start = getTime()
                try {
                    takeAndSavePhoto()
                } catch (e: Exception) {
                    println("Photo $i failed (timing): ${e.message}")
                    continue
                }
                val duration = getTime() - start
                performanceCalculator.sampleTime(duration)
                println("Saved photo ${i + 1}/$n")
            }
            performanceCalculator.postTimes()
        } else {
            performanceCalculator.start()
            for (i in 0 until n) {
                performanceCalculator.markIteration(i)
                try {
                    takeAndSavePhoto()
                } catch (e: Exception) {
                    println("Photo $i failed (performance): ${e.message}")
                    continue
                }
                println("Saved photo ${i + 1}/$n")
            }
            performanceCalculator.stopAndPost()
        }

        stopCameraSession()
        println("Camera benchmark complete")
    }
}
